zcash_primitives/src/note_encryption.rs

```diff
@@ -0,0 +1,1079 @@
+
+impl Default for Memo {
+    fn default() -> Self {
+        // Empty memo field indication per ZIP 302
+        let mut memo = [0u8; 512];
+        memo[0] = 0xF6;
+        Memo(memo)
+    }
+}
+
+impl PartialEq for Memo {
+    fn eq(&self, rhs: &Memo) -> bool {
+        &self.0[..] == &rhs.0[..]
+    }
+}
+
+impl Memo {
+    /// Returns a `Memo` containing the given slice, appending with zero bytes if
+    /// necessary, or `None` if the slice is too long. If the slice is empty,
+    /// `Memo::default` is returned.
+    pub fn from_bytes(memo: &[u8]) -> Option<Memo> {
+        if memo.is_empty() {
+            Some(Memo::default())
+        } else if memo.len() <= 512 {
+            let mut data = [0; 512];
+            data[0..memo.len()].copy_from_slice(memo);
+            Some(Memo(data))
+        } else {
+            // memo is too long
+            None
+        }
+    }
+
+    /// Returns a `Memo` containing the given string, or `None` if the string is too long.
+    pub fn from_str(memo: &str) -> Option<Memo> {
+        Memo::from_bytes(memo.as_bytes())
+    }
+
+    /// Returns the underlying bytes of the `Memo`.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0[..]
+    }
+
+    /// Returns:
+    /// - `None` if the memo is not text
+    /// - `Some(Ok(memo))` if the memo contains a valid UTF-8 string
+    /// - `Some(Err(e))` if the memo contains invalid UTF-8
+    pub fn to_utf8(&self) -> Option<Result<String, str::Utf8Error>> {
+        // Check if it is a text or binary memo
+        if self.0[0] < 0xF5 {
+            // Drop trailing zeroes
+            let mut data = &self.0[..];
+            while let Some((0, next)) = data.split_last() {
+                data = next;
+            }
+            // Check if it is valid UTF8
+            Some(str::from_utf8(data).map(|memo| memo.to_owned()))
+        } else {
+            None
+        }
+    }
+}
+
+fn generate_esk() -> Fs {
+    // create random 64 byte buffer
+    let mut rng = OsRng::new().expect("should be able to construct RNG");
+    let mut buffer = [0u8; 64];
+    for i in 0..buffer.len() {
+        buffer[i] = rng.gen();
+    }
+
+    // reduce to uniform value
+    Fs::to_uniform(&buffer[..])
+}
+
+/// Sapling key agreement for note encryption.
+///
+/// Implements section 5.4.4.3 of the Zcash Protocol Specification.
+pub fn sapling_ka_agree<'a, P>(esk: &Fs, pk_d: &'a P) -> [u8; 32]
+where
+    edwards::Point<Bls12, Unknown>: From<&'a P>,
+{
+    let p: edwards::Point<Bls12, Unknown> = pk_d.into();
+
+    // Multiply by 8
+    let p = p.mul_by_cofactor(&JUBJUB);
+
+    // Multiply by esk
+    let p = p.mul(*esk, &JUBJUB);
+
+    // Produce result
+    let mut result = [0; 32];
+    p.write(&mut result[..]).expect("length is not 32 bytes");
+    result
+}
+
+/// Sapling KDF for note encryption.
+///
+/// Implements section 5.4.4.4 of the Zcash Protocol Specification.
+fn kdf_sapling(dhsecret: &[u8], epk: &edwards::Point<Bls12, PrimeOrder>) -> Blake2bResult {
```

Comment:
`sapling_ka_agree()` was implemented by @gtank to return a byte array, likely because you never use the output of ECDH as a point (or directly at all, since it should always go through a KDF). But it's a fair point (no pun intended) that KA<sup>Sapling</sup>.Agree returns a point, and KDF<sup>Sapling</sup> takes two points.